crates/zksync/core/src/vm/runner.rs

```diff
@@ -269,21 +269,24 @@ where
     <DB as Database>::Error: Debug,
 {
     let value = ecx.env.tx.value.to_u256();
-    let balance = ZKVMData::new(ecx).get_balance(caller);
-    if balance.is_zero() {
-        error!("balance is 0 for {caller:?}, transaction will fail");
-    }
-    let max_fee_per_gas = fix_l2_gas_price(ecx.env.tx.gas_price.to_u256());
-
     let use_paymaster = !paymaster_params.paymaster.is_zero();
 
-    // We check if the paymaster is set, if it is not set, we use the proposed gas limit
-    let gas_limit = if use_paymaster {
-        ecx.env.tx.gas_limit.into()
+    // Get balance of either paymaster or caller depending on who's paying
+    let balance = if use_paymaster {
+        let paymaster_addr = Address::from_slice(paymaster_params.paymaster.as_bytes());
+        ZKVMData::new(ecx).get_balance(paymaster_addr)
     } else {
-        fix_l2_gas_limit(ecx.env.tx.gas_limit.into(), max_fee_per_gas, value, balance)
+        ZKVMData::new(ecx).get_balance(caller)
     };
 
+    if balance.is_zero() {
+        let address = if use_paymaster { paymaster_params.paymaster } else { caller.to_h160() };
```

Comment:
you can avoid this if you first get the `address` from either pm or caller, and then use that to get the `balance`, and reuse the `address` you obtained, and avoid having the same check twice